Initial naga changes for mesh shaders

inner-daemons opened the pull request:
**Connections**
#7197

**Description**
This PR isn't meant to be merged itself, its obviously way too large for that. Instead I will be breaking off chunks into separate PRs.

**Testing**
There is a test WGSL shader that is parsed and then written to SPIR-V

**Squash or Rebase?**
Absolutely squash

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

inner-daemons committed:
Initial(untested commit), vulkan and gles only supported

inner-daemons committed:
Maybe fixed compiles for metal and dx12

inner-daemons committed:
Merge branch 'gfx-rs:trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Hopefully fixed compiles for other backends and updated to functional(?) vulkan thing

inner-daemons committed:
I don't get git
 Please enter a commit message to explain why this merge is necessary,

inner-daemons committed:
Fixed the clippy warning

inner-daemons committed:
Initial naga changes and outline

inner-daemons committed:
Fixed minor issue

inner-daemons committed:
Merge recent changes, most significantly features type overhaul

inner-daemons committed:
Merge branch 'mesh-shading/wgpu-hal' into mesh-shading/naga-init

inner-daemons committed:
Fixed silly documentation mistake

inner-daemons committed:
Merge branch 'mesh-shading/wgpu-hal' into mesh-shading/naga-init

inner-daemons committed:
Hopefully fixed some issues in CI

inner-daemons committed:
Automatic changes caused by tests

inner-daemons committed:
Fixed issue with multiview feature

inner-daemons committed:
Merge branch 'mesh-shading/wgpu-hal' into mesh-shading/naga-init

inner-daemons committed:
CI will hate this one - made some changes based on discussion of proposal, left a lot of todo!()'s

inner-daemons committed:
Made CI slightly less angry

inner-daemons committed:
Dummy commit for dummy CI

The CI pooped itself, hopefully this fixes that. Will probably be undone either way.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/wgpu-hal

inner-daemons committed:
Merge branch 'mesh-shading/wgpu-hal' into mesh-shading/naga-init

inner-daemons committed:
(Tried to) Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Tried to undo cargo.lock changes

inner-daemons committed:
Dusted off this ancient PR

inner-daemons committed:
Updated naga snapshots

inner-daemons committed:
A little more work

inner-daemons committed:
Worked a little more

inner-daemons committed:
Now (almost) parses!

inner-daemons committed:
Updated wgsl file to not specify unsigned integer literals. Will work on auto interpreting that later

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Added more stuff to naga IR and a tiny bit of validation

inner-daemons committed:
Fixed some typos lol

inner-daemons committed:
Added changelog entry

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Added task_payload storage class parsing to wgsl, validation for it, etc

inner-daemons committed:
Added output type validation

inner-daemons committed:
Tried to fix u32 casting

inner-daemons committed:
Attempted to fix u32 to i32 automatic conversion issue

inner-daemons committed:
Fixed stupid naga validation bug

inner-daemons committed:
Updated snapshots

inner-daemons committed:
Tried to clarify mesh validation errors

inner-daemons committed:
Made it adjust mesh shader metadata during compaction

inner-daemons committed:
Added more validation for the output type of entry points

inner-daemons committed:
Made it allow @mesh_task_size declared in a struct for task output

inner-daemons committed:
Smartified the task output checking

inner-daemons commented:
@cwfitzgerald There are some correctness checks that I suspect would be hard to implement. For example, `setMeshOutputs` must be called before exiting and before any call to `setVertex` or `setPrimitive`, but can only be called once. So if we can't verify that `setMeshOutputs` will be called before other functions, we can't just insert an empty call. We also have to ensure that it doesn't get called multiple times.

I'm thinking that we should do advanced function analysis that will inevitably miss some valid code, and then add a flag so that creating a trusted shader skips this. Does that make sense to you?

Edit: also probably relevant to this question is bounds checking, which would depend on the output of `setMeshOutputs`.

inner-daemons committed:
A little more work for SPIRV writing and added back point index builtin

inner-daemons committed:
Now it can (almost) compile a task shader!

inner-daemons committed:
Task shaders can now be written properly!!!!

inner-daemons committed:
Cleaned up task shader writing

inner-daemons committed:
More minor fixes to task shader and made it require SPIR-V 1.4

inner-daemons committed:
Now also passes spirv-val

inner-daemons committed:
Fixed silly typo that CI noticed

inner-daemons committed:
Updated snapshots

inner-daemons committed:
Attempted to make a small adjustment to SPIR-V writing

inner-daemons committed:
A lot more progress made

inner-daemons committed:
HAHAHAH YEASSS

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Updated snapshots

inner-daemons committed:
Updated mesh shading spec, also to include @per_primitive attribute

inner-daemons committed:
Final tweaks to the spec

inner-daemons committed:
Small tweaks

inner-daemons committed:
A few more small tweaks

inner-daemons committed:
More nitpicky tweaks, next focus is actually addressing the problem in spirv writing

inner-daemons committed:
Pushing broken changes for now because I have to switch dev platforms due to constant crashes

inner-daemons committed:
Updated to not have 2 arrays of structs (gross)

inner-daemons committed:
Updated snapshots

inner-daemons committed:
Tried to hopefully fix stupid issue

inner-daemons committed:
Updated mesh-shader.toml config

inner-daemons committed:
Update rspirv version

inner-daemons committed:
Updated naga snapshots

inner-daemons committed:
Merge branch 'update-rspirv-version' into mesh-shading/naga-init

inner-daemons committed:
Updated snapshots

inner-daemons committed:
Reformatted the config toml

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons commented:
I'm marking this as ready for review so that it can generate some eyes to look at the general process. Obviously the PR is 5k lines long so will not be merged in anything like its current state. It also is failing CI due to rspirv not being updated (see #7945 ) and the shader compiling benchmark not respecting the configuration tomls and therefore trying to compile for incompatible backends like GLSL.

The fragment shader input on the test would be invalid to use with the corresponding mesh shader because it doesn't have the not yet implemented `@per_primitive` attribute for its primitive input.

I'd mainly like for people to check out my SPIRV backend changes, as that is the part I am most unsure about.

The generated mesh and task shaders pass spirv-val, but I haven't yet tried using them in a real mesh pipeline.

inner-daemons committed:
Implemented coordinate flipping

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons commented:
Unmarking as draft (again lol) because I have given it another pass with my own eyes and @cwfitzgerald mentioned he would check it out

Also I'd like to know how you'd prefer to address the benchmark issue

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

jimblandy commented:
@SupaMaggie70Incorporated Would you like to rebase this and address the CI failures?

inner-daemons commented:
@jimblandy Rebase onto what exactly? Also, the CI failures are due to shader compiling benchmarks that can't be told "only compile this into SPIRV" since not all output languages are supported yet. So I am actually looking for guidance on how I should address that!

This is marked as a non-draft despite its size because @cwfitzgerald mentioned he wanted to take a look over the whole thing before I split it up and let other people review individual PRs. The PR isn't *quite* done as is, but I just wanted to give him the chance to look over approximately what it would be.

andyleiserson commented:
> Also, the CI failures are due to shader compiling benchmarks that can't be told "only compile this into SPIRV" since not all output languages are supported yet. So I am actually looking for guidance on how I should address that!

There are toml files alongside the test shaders that specify which backends they are intended for, although I'm not sure if what's there will match exactly with what the benchmark test needs. But an option to consider is figuring out how to reuse the code in `naga/tests/naga/snapshots.rs` that parses these toml files for the benchmark.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Updated shader test outputs

inner-daemons committed:
Added per primitive stuff

inner-daemons committed:
Tried to completely overhaul the output variable writing for spirv

inner-daemons committed:
Committing this nonsense before it all gets destroyed again by my stupid fucking pc

inner-daemons committed:
A little more progress (I'm paranoid)

inner-daemons committed:
A little more progress (still very broken)

inner-daemons committed:
Updated my todos

inner-daemons committed:
It frickin works now I think

inner-daemons committed:
Implemented coordinate flip

inner-daemons committed:
Made the example properly work with ALL naga shaders!!

inner-daemons committed:
Upgraded the copy algorithm to not be complete garbage

inner-daemons committed:
Final tweaks and removing // TODO 's

inner-daemons commented:
@andyleiserson Preferably I would want the toml configuration logic to be shared and not copy/pasted between different crates. But I am not familiar enough with the project structure to know where it should live, or if the naga benchmarks should be moved to the crate in naga/tests/naga. So if you or someone else could provide an opinion about how to handle that I'd be happy to do the leg work (probably in a separate PR).

That being said, I get the impression that refactors of code/moving code around should only be done by trusted people, as other individuals could hide something malicious in the large change. Not sure about wgpu's process in that regard but something to keep in mind probably.

cwfitzgerald commented on naga/src/back/spv/block.rs:
```diff
@@ -253,6 +257,349 @@ impl Writer {
                 _ => {}
             }
         }
+        // MeshTaskSize must be called write before exiting
```
Typo maybe? This comment doesn't make sense

cwfitzgerald commented on naga/src/back/spv/block.rs:
```diff
@@ -253,6 +257,349 @@ impl Writer {
+        return_info: &super::MeshReturnInfo,
+        body: &mut Vec<Instruction>,
+    ) -> Result<(), Error> {
+        let vert_info = self.mesh_shader_output_variable(
```
This could use a number of comments explaining what we're doing here and why we need to do it for mesh shaders. This is a very long and important function, so comments throughout explaining it all (will also help).

inner-daemons replied:
I attempted to write some comments. Let me know if you think I should do more.

cwfitzgerald commented on naga/src/back/spv/block.rs:
```diff
@@ -3626,6 +3979,117 @@ impl BlockContext<'_> {
+                            && matches!(
+                                member_info.member.binding,
+                                Some(crate::Binding::BuiltIn(crate::BuiltIn::Position { .. }))
+                            )
+                        {
+                            let float_ptr_type_id = self
+                                .writer
+                                .get_f32_pointer_type_id(spirv::StorageClass::Output);
+                            let index_y_id = self.get_index_constant(1);
+
+                            let access_id = self.gen_id();
+                            block.body.push(Instruction::access_chain(
+                                float_ptr_type_id,
+                                access_id,
+                                out_ptr_id,
+                                &[index_y_id],
+                            ));
+
+                            let float_type_id = self.writer.get_f32_type_id();
+                            let y_val = self.gen_id();
+                            block.body.push(Instruction::composite_extract(
+                                float_type_id,
+                                y_val,
+                                in_value_id,
+                                &[1],
+                            ));
+
+                            let flipped_y_val = self.gen_id();
+                            block.body.push(Instruction::unary(
+                                spirv::Op::FNegate,
+                                float_type_id,
+                                flipped_y_val,
+                                y_val,
+                            ));
+
+                            block
+                                .body
+                                .push(Instruction::store(access_id, flipped_y_val, None));
+                        }
+                    }*/
```
This commented block should be removed.

cwfitzgerald commented on naga/src/back/spv/writer.rs:
```diff
@@ -833,15 +893,22 @@ impl Writer {
+                            let varying_id = if *binding
+                                == crate::Binding::BuiltIn(crate::BuiltIn::MeshTaskSize)
+                            {
+                                0
```
This feels very easy to create a bug with? Are these rewritten later or something?

inner-daemons replied:
I updated to add a comment, but basically MeshTaskSize isn't an actual builtin but rather something that needs to be a parameter to a final terminator, which is handled elsewhere

And varying_id is ignored in the output writing in this case.

cwfitzgerald commented on naga/src/back/spv/writer.rs:
```diff
@@ -881,6 +948,357 @@ impl Writer {
             None => self.void_type,
         };
 
+        if let Some(ref mut iface) = interface {
```
Could any of this massive block be split out to another function or something?

inner-daemons replied:
Done!

cwfitzgerald commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -3130,6 +3232,60 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+
+                            let rctx = ctx.runtime_expression_ctx(span)?;
+
+                            // TODO: fix this
```
Does this need to be addressed?

inner-daemons replied:
Nope

cwfitzgerald commented on naga/src/ir/mod.rs:
```diff
@@ -329,6 +329,16 @@ pub enum ShaderStage {
 }
 
+impl ShaderStage {
+    // TODO: make more things respect this
```
Does this need to be addressed?

inner-daemons replied:
Done!

cwfitzgerald commented on naga/src/ir/mod.rs:
```diff
@@ -2578,3 +2603,47 @@ pub struct Module {
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[allow(dead_code)]
+pub struct MeshStageInfo {
+    /// Should be Some by the time it passes validation
```
Outdated comment, this isn't optional

cwfitzgerald reviewed (changes requested):
This is an absolutely herculean effort! I didn't see much that is would be of major concern.

I would advise however splitting up the PR, either into multiple PRs or multiple independent commits. A possible suggestion for splitting

- IR changes + ron test output changes
- WGSL frontend
- Spirv backend

The wgsl frontend and spirv backend could be completely independent PRs, both dependent on the fairly minor IR changes. I think that would put diff sizes in the more reasonable range.

inner-daemons committed:
Tried to address some comments and also add comments

inner-daemons committed:
Addressed another comment

inner-daemons commented:
@cwfitzgerald Thanks for the review. I've tried to sprinkle a few more comments around the code in addition to addressing your comments, so if you could ensure that the code is understandable to yourself and others that would be great. Your idea for splitting up the PR is pretty much exactly what I was thinking. I will probably break off the naga IR changes tomorrow. I'll also hopefully find the motivation to fix the benchmark issue before WGSL parsing gets merged in.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Quick before corruption!

inner-daemons committed:
A few more changes

inner-daemons committed:
Tried to fix a stupid accidentally pushed issue

inner-daemons committed:
Brief little fix for bad paths

inner-daemons committed:
Removed `const` thing from a functoin that cant be const on older compilers

inner-daemons committed:
Broken shaders now break :party:

inner-daemons committed:
re-unbroke broken shaders

inner-daemons committed:
Fixed compile fail failing

inner-daemons committed:
Tried to fix CI failing on

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-init

inner-daemons committed:
Updated snapshots

cwfitzgerald commented:
I'm going to close this now that it is on its way to be splitten up.
